### PR TITLE
Direct solvers on GPU

### PR DESCRIPTION
**Pull request by Rombur**
This adds direct solvers on the GPUs. There is a lot of code which at first looks duplicated but it's only because the functions in `cusolver` have a slightly different name for `float` and `double`. 

@tamiko can you take a look at the (horrible) change I've made to `cmake/modules/FindCUDA.cmake` This actually kinda works but for one of the solver I get this error: `undefined symbol: omp_get_max_thread` which it's probably due to the way I've added OpenMP. If what I did looks alright, I will just remove the failing solver since the two others work fine.

### CODE REVIEW

**Comment by Rombur** [comment]
_ping_

File: include/deal.II/lac/cuda_solver_direct.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace CUDAWrappers
+{
+  /**
+   * Direct solvers. These solvers call cuSOLVER underneath.
+   *
+   * @note Instantiations for this template are provided for <tt>@<float@> and
+   * @<double@></tt>.
```
Comment (kronbichler) [inline include/deal.II/lac/cuda_solver_direct.h]: Nit-pick: Would you mind moving the `and` out of the `tt` style?

File: include/deal.II/lac/cuda_solver_direct.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      /**
+       * Set the additional data field to the desired solver.
+       */
+      explicit
+      AdditionalData (const std::string &solver_type = "LU_dense");
+
+      /**
+       * Set the solver type. Possibilities are:
+       * <ul>
+       * <li> "Cholesky" which uses Cholesky on the device </li>
```
Comment (kronbichler) [inline include/deal.II/lac/cuda_solver_direct.h]: How about: "Cholesky which performs a Cholesky decomposition on the device"?

File: include/deal.II/lac/cuda_solver_direct.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+       * <li> "LU_dense" which converts the sparse matrix to a dense matrix and
+       * uses LU factorization </li>
+       * <li> "LU_host" which uses LU factorization on the host </li>
```
Comment (kronbichler) [inline include/deal.II/lac/cuda_solver_direct.h]: Out of curiosity, and before reading the full PR to the end: What does "LU factorization on the host" mean technically? Does CUDA include a sparse LU factorization package? Is it one of the well-known ones like Pardiso or UMFPACK?

File: include/deal.II/lac/cuda_solver_direct.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+     * Destructor.
+     */
+    virtual ~SolverDirect() = default;
+
+    /**
+     * Solve the linear system <tt>Ax=b</tt>.
+     */
+    void solve(const SparseMatrix<Number> &A,
+               LinearAlgebra::CUDAWrappers::Vector<Number> &x,
+               const LinearAlgebra::CUDAWrappers::Vector<Number> &b);
```
Comment (kronbichler) [inline include/deal.II/lac/cuda_solver_direct.h]: This is a nice interface and in line with our other classes. :+1:

File: source/base/exceptions.cc [hunk 5]
```diff
@@ -522,6 +522,48 @@ namespace deal_II_exceptions
...
+        {
+          return "The cuSolver library was not initialized";
+        }
+        case CUSOLVER_STATUS_ALLOC_FAILED:
+        {
+          return "Resource allocation failed inside the cuSolver library";
+        }
+        case CUSOLVER_STATUS_INVALID_VALUE:
+        {
+          return "An unsupported value of parameter was passed to the function";
```
Comment (kronbichler) [inline source/base/exceptions.cc]: of A parameter

File: source/base/exceptions.cc [hunk 6]
```diff
@@ -522,6 +522,48 @@ namespace deal_II_exceptions
...
+        }
+        case CUSOLVER_STATUS_ARCH_MISMATCH:
+        {
+          return "The function requires a feature absent from the device architecture";
```
Comment (kronbichler) [inline source/base/exceptions.cc]: Can we easily query the SM architecture here to add what the compiled version is? I know that one sets these in the makefile, but I have not checked if it is possible to query them at compile time. Or is this error message a hardware error not related to the compilation at all, so giving this info would not be useful at all? If I remember correctly, I have chosen different SM modes on Pascal at compile time.

File: source/lac/cuda_solver_direct.cu [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    {
+      auto cusparse_matrix = matrix.get_cusparse_matrix();
+
+      cusparseStatus_t cusparse_error_code =
+        cusparseScsr2dense(
+          cusparse_handle, matrix.m(), matrix.n(),
+          std::get<3>(cusparse_matrix), std::get<0>(cusparse_matrix),
+          std::get<2>(cusparse_matrix), std::get<1>(cusparse_matrix),
+          dense_matrix_dev, matrix.m());
+      AssertCusparse(cusparse_error_code);
```
Comment (kronbichler) [inline source/lac/cuda_solver_direct.cu]: Just to help me understand the calling structure: This call is issued on the host but uses pointers on the device, so the library function `cusparseScsr2dense` does the operations on the host?

File: source/lac/cuda_solver_direct.cu [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+    void cusolverDngetrs(cusolverDnHandle_t cusolver_dn_handle, int m,
+                         double *dense_matrix_dev, int *pivot_dev, double *b,
+                         int *info_dev)
+    {
+      const int n_rhs = 1;
+      cusolverStatus_t cusolver_error_code =
+        cusolverDnDgetrs(cusolver_dn_handle, CUBLAS_OP_N, m, n_rhs,
+                         dense_matrix_dev, m, pivot_dev, b, m, info_dev);
+      AssertCusolver(cusolver_error_code);
```
Comment (kronbichler) [inline source/lac/cuda_solver_direct.cu]: It's a pity that the cuda libraries (or LAPACK to begin with) are not templated and you have to list all those functions twice... :grinning:

File: source/lac/cuda_solver_direct.cu [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+    }
+
+
+
+    void cusolverSpcsrlsvluHost(cusolverSpHandle_t cusolver_sp_handle,
+                                const unsigned int n_rows, unsigned int nnz,
+                                cusparseMatDescr_t descr, const double *val_host,
+                                const int *row_ptr_host,
+                                const int *column_index_host, const double *b_host,
+                                double *x_host)
```
Comment (kronbichler) [inline source/lac/cuda_solver_direct.cu]: To ease reading, could you please break the lines in function arguments at the same place as in the `float` case?

File: source/lac/cuda_solver_direct.cu [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+    template <typename Number>
+    void lu_factorization(cusolverSpHandle_t cusolver_sp_handle,
+                          const SparseMatrix<Number> &matrix,
+                          const Number *b_dev, Number *x_dev)
+    {
+      // cuSOLVER does not support LU factorization of sparse matrix on the device,
+      // so we need to move everything to the host first and then back to the host.
+      const unsigned int nnz = matrix.n_nonzero_elements();
+      const unsigned int n_rows = matrix.m();
```
Comment (kronbichler) [inline source/lac/cuda_solver_direct.cu]: Again, to help me understand the model here: `b_dev` and `x_dev` are both pointers in device memory, and `matrix` is also (by design) on the device, right?

File: include/deal.II/lac/cuda_solver_direct.h (hunk 3 above)
Comment (Rombur) [reply include/deal.II/lac/cuda_solver_direct.h]: This means that the data is moved to the host where the factorization is done. Unfortunately, there is no function to do a LU decomposition on the device if the matrix is sparse. Like the two other functions this function uses `cuSOLVER`.

File: source/base/exceptions.cc (hunk 6 above)
Comment (Rombur) [reply source/base/exceptions.cc]: I am not sure what you mean here. The compilation in cuda is really in two phases: in the first phase, you decide which capability you want and in the second phase, you compile for the architecture. This allows you to compile with Kepler capabilities but run on Volta. I am not sure what would happen if you decide to compile for Fermi and run Volta but this error should never be triggered in practice. According to the documentation, this only happens for compute capability below 2.0 and we require a compute capability of 3.5. I've just added the error for completeness.

File: source/lac/cuda_solver_direct.cu (hunk 7 above)
Comment (Rombur) [reply source/lac/cuda_solver_direct.cu]: Almost right :smile: The call is on the host, the pointers are on the device, and (presumably) `cusparseScsr2dense` is on the device. Maybe internally this is done on the host and then moved back to the device but there is really no way to know. To do something on the host, the pointers should be on the host, and the function would be called something like `cusparseScsr2denseHost`

File: source/lac/cuda_solver_direct.cu (hunk 10 above)
Comment (Rombur) [reply source/lac/cuda_solver_direct.cu]: Yes `b_dev` and `x_dev` are both pointers in device memory. `matrix` is our own object which has all of its data on the device.

File: source/lac/cuda_solver_direct.cu (hunk 7 above)
Comment (kronbichler) [reply source/lac/cuda_solver_direct.cu]: Thanks, that makes sense.

File: source/base/exceptions.cc (hunk 6 above)
Comment (kronbichler) [reply source/base/exceptions.cc]: Thanks for the clarification, your answer is precisely telling me what I wanted to know. :+1:

File: source/lac/cuda_solver_direct.cu [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+
+
+
+    void cusparsecsr2dense(cusparseHandle_t cusparse_handle,
+                           const SparseMatrix<double> &matrix,
+                           double *dense_matrix_dev)
+    {
+      auto cusparse_matrix = matrix.get_cusparse_matrix();
+
+      cusparseStatus_t cusparse_error_code =
```
Comment (bangerth) [inline source/lac/cuda_solver_direct.cu]: here and in other places of the file, could you make it `const`?

File: source/lac/cuda_solver_direct.cu [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+
+#include <deal.II/lac/cuda_solver_direct.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace CUDAWrappers
+{
+  namespace internal
+  {
+    void cusparsecsr2dense(cusparseHandle_t cusparse_handle,
```
Comment (bangerth) [inline source/lac/cuda_solver_direct.cu]: I guess these are all internal functions you never intend to call from any other .cc file, right? If so, you could wrap them in another level of anonymous namespace so their names don't pollute the linker space.

**Pull request merged**